src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs

```diff
@@ -14,79 +14,257 @@ namespace Microsoft.CodeAnalysis.CSharp
-        internal static readonly Conversion ImplicitEnumeration = new Conversion(ConversionKind.ImplicitEnumeration);
-        internal static readonly Conversion ImplicitTupleLiteral = new Conversion(ConversionKind.ImplicitTupleLiteral);
-        internal static readonly Conversion ImplicitTuple = new Conversion(ConversionKind.ImplicitTuple);
-        internal static readonly Conversion ExplicitTuple = new Conversion(ConversionKind.ExplicitTuple);
-        internal static readonly Conversion AnonymousFunction = new Conversion(ConversionKind.AnonymousFunction);
-        internal static readonly Conversion Boxing = new Conversion(ConversionKind.Boxing);
-        internal static readonly Conversion NullLiteral = new Conversion(ConversionKind.NullLiteral);
-        internal static readonly Conversion NullToPointer = new Conversion(ConversionKind.NullToPointer);
-        internal static readonly Conversion PointerToVoid = new Conversion(ConversionKind.PointerToVoid);
-        internal static readonly Conversion PointerToPointer = new Conversion(ConversionKind.PointerToPointer);
-        internal static readonly Conversion PointerToInteger = new Conversion(ConversionKind.PointerToInteger);
-        internal static readonly Conversion IntegerToPointer = new Conversion(ConversionKind.IntegerToPointer);
-        internal static readonly Conversion Unboxing = new Conversion(ConversionKind.Unboxing);
-        internal static readonly Conversion ExplicitReference = new Conversion(ConversionKind.ExplicitReference);
-        internal static readonly Conversion IntPtr = new Conversion(ConversionKind.IntPtr);
-        internal static readonly Conversion ExplicitEnumeration = new Conversion(ConversionKind.ExplicitEnumeration);
-        internal static readonly Conversion ExplicitNullable = new Conversion(ConversionKind.ExplicitNullable);
-        internal static readonly Conversion ExplicitNumeric = new Conversion(ConversionKind.ExplicitNumeric);
-        internal static readonly Conversion ImplicitDynamic = new Conversion(ConversionKind.ImplicitDynamic);
-        internal static readonly Conversion ExplicitDynamic = new Conversion(ConversionKind.ExplicitDynamic);
-        internal static readonly Conversion InterpolatedString = new Conversion(ConversionKind.InterpolatedString);
-
-        private readonly MethodSymbol _methodGroupConversionMethod;
-
-        private readonly UserDefinedConversionResult _conversionResult; //no effect on Equals/GetHashCode
-
         private readonly ConversionKind _kind;
-        private readonly byte _flags;
+        private readonly UncommonData _uncommonData;
+
+        // most conversions are trivial and do not require additional data besides Kind
+        // in uncommon cases an instance of this class is attached to the conversion.
+        private class UncommonData
+        {
+            public UncommonData(
```

Comment:
This API exists for performance reasons and is a bit unsafe. In many cases you would need to pass some additional data besides kind and accidental omitting some data could lead to subtly incorrect behavior.

I was concerned that with default values it could be too easy to pass incomplete set of data, so I want to force the caller to pass default values explicitly (preferable using named parameters) and think/comment why defaults are acceptable.

It should not be very common to call this directly, so I think it is not a big burden.